### PR TITLE
Added offline editing WFS tests and debug info

### PR DESCRIPTION
**Pull request by elpaso**

### CODE REVIEW

File: tests/src/python/test_qgsserver_offline_editing.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        self.assertTrue(offline_layer.changeAttributeValue(feat2.id(), offline_layer.fieldNameIndex('name'), 'name 2 edited'))
+        self.assertTrue(offline_layer.changeGeometry(feat2.id(), QgsGeometry.fromPoint(QgsPoint(33.0, 60.0))))
+        self.assertTrue(offline_layer.commitChanges())
+        feat2 = self._getFeatureByAttribute(offline_layer, 'name', "'name 2 edited'")
+        self.assertTrue(ol.isOfflineProject())
+        self.assertTrue(self.project.readEntry("OfflineEditingPlugin", "/OfflineDbPath")[1])
+        # Sync
+        ol.synchronize()
+        # Does anybody knows why the sleep is needed? Is that a threaded WFS consequence?
+        sleep(1)
```
Comment (elpaso) [inline tests/src/python/test_qgsserver_offline_editing.py]: @rouault do you have any clue why a sleep is required here to get WFS modified features?

File: src/core/qgsofflineediting.cpp [hunk 2]
```diff
@@ -817,7 +831,7 @@ void QgsOfflineEditing::applyAttributeValueChanges( QgsVectorLayer* offlineLayer
   for ( int i = 0; i < values.size(); i++ )
   {
     QgsFeatureId fid = remoteFid( db, layerId, values.at( i ).fid );
-
+    QgsDebugMsg( QString( "Offline changeAttributeValue %1 = %2" ).arg( QString( attrLookup[ values.at( i ).attr ] ), values.at( i ).value ) );
```
Comment (m-kuhn) [inline src/core/qgsofflineediting.cpp]: Can you make this a `QgsDebugMsgLevel` call?

File: tests/src/python/test_qgsserver_offline_editing.py (hunk 1 above)
Comment (m-kuhn) [inline tests/src/python/test_qgsserver_offline_editing.py]: Maybe the WFS provider works asynchroneously? Or the server?

Anyway, I was just right now writing tests for this, but if you do it I'll stop .

File: src/core/qgsofflineediting.cpp (hunk 2 above)
Comment (m-kuhn) [inline src/core/qgsofflineediting.cpp]: Or remove it completely ;)

File: tests/src/python/test_qgsserver_offline_editing.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    QgsProject,
+    QgsMapLayerRegistry,
+    QgsOfflineEditing,
+)
+from qgis.testing import (
+    start_app,
+    unittest,
+)
+
+from PyQt.QtCore import QFileInfo
```
Comment (m-kuhn) [inline tests/src/python/test_qgsserver_offline_editing.py]: Should be `qgis.PyQt.QtCore`

File: tests/src/python/test_qgsserver_offline_editing.py (hunk 1 above)
Comment (elpaso) [inline tests/src/python/test_qgsserver_offline_editing.py]: We could join efforts :)
You may add the tests for postgis?

File: src/core/qgsofflineediting.cpp (hunk 2 above)
Comment (elpaso) [inline src/core/qgsofflineediting.cpp]: Yes, I'd keep it for a while until things are fixed then remove all debug stuff.

File: tests/src/python/test_qgsserver_offline_editing.py (hunk 1 above)
Comment (m-kuhn) [inline tests/src/python/test_qgsserver_offline_editing.py]: That was the idea

File: src/core/qgsofflineediting.cpp (hunk 2 above)
Comment (m-kuhn) [inline src/core/qgsofflineediting.cpp]: Is there something left to fix?
This call here is probably overly verbose in a real life scenario.
And some others could probably directly go to the message log and make the user aware of a potential problem (when there are no debug builds)

**Comment by m-kuhn** [comment]
Is this related to qgsserver (Test name)?

File: tests/src/python/test_qgsserver_offline_editing.py (hunk 1 above)
Comment (rouault) [inline tests/src/python/test_qgsserver_offline_editing.py]: Hum, no definite idea to explain the need for the sleep. I can't think to a reason on the client side (but I can be wrong). A potential explanation would be that the server acknowledges the WFS-T request to the client, but would commit the changes in a separate thread, but I doubt it is architectured like that (I haven't checked).

File: src/core/qgsofflineediting.cpp (hunk 2 above)
Comment (elpaso) [inline src/core/qgsofflineediting.cpp]: No problem to remove that debug message.
Do you mean that the other debug messages should become `QgsDebugMsgLevel` ? What level do you think would be appropriate?

I'm not sure that thing are really fixed for WFS, this test passes locally but I'd like to add another test for the case:
- go offline
- edit
- delete the first or the last feature with an online WFS call
- sync

I fear that the fid<->id mapping will non be correct anymore, don't we really need a WFS fid<->id mapping stored in the offline DB?

Also, tests cases for offline INSERT and DELETE are still missing.

**Comment by elpaso** [comment]
The test comes from the fact that it uses QGIS server as WFS-T endpoint.
If you are going to add PostGIS tests here, I'd probably change it to just `test_offline_editing.py`

File: src/core/qgsofflineediting.cpp (hunk 2 above)
Comment (m-kuhn) [inline src/core/qgsofflineediting.cpp]: On 06/30/2016 09:55 AM, Alessandro Pasotti wrote:
> In src/core/qgsofflineediting.cpp
> <https://github.com/qgis/QGIS/pull/3258#discussion_r69086725>:
> 
>> @@ -817,7 +831,7 @@ void QgsOfflineEditing::applyAttributeValueChanges( QgsVectorLayer* offlineLayer
>>    for ( int i = 0; i < values.size(); i++ )
>>    {
>>      QgsFeatureId fid = remoteFid( db, layerId, values.at( i ).fid );
>> -
>> +    QgsDebugMsg( QString( "Offline changeAttributeValue %1 = %2" ).arg( QString( attrLookup[ values.at( i ).attr ] ), values.at( i ).value ) );
> 
> No problem to remove that debug message.
> Do you mean that the other debug messages should become
> |QgsDebugMsgLevel| ? What level do you think would be appropriate?

I always use 4 for stuff that I really only need for debugging.
Except for the ones that should actually never trigger (and depict real
errors) which don't need the level and could possibly go to
QgsMessageLog (and end up in the user interface)

> 
> I'm not sure that thing are really fixed for WFS, this test passes
> locally but I'd like to add another test for the case:
> 
>   * go offline
>   * edit
>   * delete the first or the last feature with an online WFS call
>   * sync
> 
> I fear that the fid<->id mapping will non be correct anymore, don't we
> really need a WFS fid<->id mapping stored in the offline DB?

Yeah, that would be a good idea.

I guess WFS fids are not stable, I don't know if there is a way to get
something unique and stable (pk) yet.

**Comment by m-kuhn** [comment]
If we add two backends I think we could make it something like the provider tests where we define the tests in a base file and then write drivers for backends that run the base tests.

**Comment by m-kuhn** [comment]
What's wrong with python 3?

**Comment by m-kuhn** [comment]
I'll have a look at the base class when I find some time.

Trying to debug a crashing WMS server now... Is there any way to do this efficiently?

**Comment by elpaso** [comment]
See: https://travis-ci.org/qgis/QGIS/jobs/141330460#L1228 and 
https://travis-ci.org/qgis/QGIS/jobs/141330460#L1228
The thing is that I don't have a QT5 Py3 build environment yet so I cannot debug it.

**Comment by m-kuhn** [comment]
You can run

    ./scripts/2to3 -w tests/src/python/qgis_wrapped_server.py
    ./scripts/2to3 -w tests/src/python/test_offline_editing_wfs.py

And hope that it helps

**Comment by m-kuhn** [comment]
How are the tests doing?
I updated osgeo4travis with spatialite support in gdal, is that working now?

**Comment by elpaso** [comment]
Tests are passing (well, were passing: before latest commits for docs coverage), merged into master: https://travis-ci.org/qgis/QGIS/builds/142147825

Next step will be to write new (failing) test cases...

**Pull request merged**